quickwit/quickwit-ingest/src/ingest_v2/workbench.rs

```diff
@@ -17,22 +17,108 @@
 
+#[derive(Default)]
+struct PublishState {
+    awaiting_publish: HashMap<ShardId, Position>,
+    already_published: HashMap<ShardId, Position>,
+}
+
+/// A helper for tracking the progress of the publish events when running in
+/// `wait_for` commit mode.
+///
+/// Registers a set of shard positions and listens to [`ShardPositionsUpdate`]
+/// events to assert when all the persisted events have been published. To make
+/// sure that no events are missed:
+/// - the tracker should be created before the persist requests are sent
+/// - the `shard_persisted` method should for all successful persist subrequests
+struct PublishTracker {
+    state: Arc<Mutex<PublishState>>,
+    publish_complete: Arc<Notify>,
+    _publish_listen_handle: EventSubscriptionHandle,
+}
+
+impl PublishTracker {
+    fn new(event_tracker: EventBroker) -> Self {
+        let state = Arc::new(Mutex::new(PublishState::default()));
+        let state_clone = state.clone();
+        let publish_complete = Arc::new(Notify::new());
+        let publish_complete_notifier = publish_complete.clone();
+        let _publish_listen_handle =
+            event_tracker.subscribe(move |update: ShardPositionsUpdate| {
+                let mut state_handle = state_clone.lock().unwrap();
+                for (updated_shard_id, updated_position) in &update.updated_shard_positions {
+                    if let Some(shard_position) =
+                        state_handle.awaiting_publish.get(updated_shard_id)
+                    {
+                        if updated_position >= shard_position {
+                            state_handle.awaiting_publish.remove(updated_shard_id);
+                            if state_handle.awaiting_publish.is_empty() {
+                                publish_complete_notifier.notify_one();
+                            }
+                        }
+                    } else {
+                        // Save this position update in case the publish update
+                        // event arrived before the shard persist response. We
+                        // might build a state that tracks irrelevant shards for
+                        // the duration of the query but that should be fine.
+                        state_handle
+                            .already_published
+                            .insert(updated_shard_id.clone(), updated_position.clone());
+                    }
+                }
+            });
+        Self {
+            state,
+            _publish_listen_handle,
+            publish_complete,
+        }
+    }
+
+    fn shard_persisted(&self, shard_id: ShardId, new_position: Position) {
+        let mut state_handle = self.state.lock().unwrap();
+        match state_handle.already_published.get(&shard_id) {
+            Some(already_published_position) if new_position <= *already_published_position => {
+                // already published, no need to track this shard's position updates
```

Comment:
I think this is a bug...
Who will be notify everyone if that happens?

Reply:
If I understand correctly, the sequence you are flagging here as a potential bug is the following:
1. the publish event is recorded
2. the persisted shard position is recorded

In that case, the shard position is never recorded in `awaiting_publish`. You are right that this means that the notification is never tirggered. But this situation is handled in the waiting method:
https://github.com/quickwit-oss/quickwit/blob/fbbdc0f442738692ded82841c9764b747bf72194/quickwit/quickwit-ingest/src/ingest_v2/workbench.rs#L111-L119

Do you think I missed something in the correctness argument?

Reply:
actually, after some extra thoughts, there is a bug if:
1. a persisted position is recorded for shard1
2. the publish event for shard1 is recorded
3. at this point the notifications is triggered and stored (notify_one)
4. a persisted shard position is recorded
5. `wait_publish_complete` is called and doesn't block because the notification was recorded